enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,177 @@
+CLI-manager will replace the current consoleclidownloads mechanism and will extend its function.
+
+2.  Central repository where all artifacts are stored - [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+currently is where crc, oc, ocp-dev-preview, ocp, odo, and serverless artifacts are published.  This is currently how OpenShift CLIs are published.  The ConsoleCLIDownloads
+CRs reference this index.  The challenge is offering these in disconnected environments and automating the download and extraction of the artifacts.  
+
+3.  Images for each CLI. The CLI in-cluster manager would provide an extract mechanism for each CLI image, similar to how we currently 
+`oc adm release extract --command oc` and `oc image extract`.   A Custom Resource for each CLI would provide the in-cluster CLI manager information about each CLI and its image.
+In disconnected, the images will be available through a mirrored local registry.  The logic for `oc image extract` can be re-used and extended for extracting the CLIs. 
+    * CLI owners build images to package artifacts similar to https://github.com/sallyom/clustercli/tree/master/hack/cli-images
+    * CLIs would be managed by users through extending the oc commands we currently have to extract `oc` and `openshift-install` binaries from the release payload.
+    This would provide the function we need, but looking to the future there is an upstream effort we can use to provide an in-cluster CLI manager while also providing 
+    a mechanism that the Kubernetes community as a whole can utilize.  That brings us to option 4.
+
+4.  The CLI in-cluster manager can provide CLIs as plugins through [krew](https://github.com/kubernetes-sigs/krew).  CLI images will serve the artifacts to enable disconnected downloads, and OpenShift CLIs will be available as 
+[krew plugins](https://github.com/kubernetes-sigs/krew-index/tree/master/plugins).
+    * Krew and Krew plugins are upstream projects that Kubernetes users are already familiar with
+    * From the [README](https://github.com/kubernetes-sigs/krew/#what-does-krew-do): `Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is similar to tools like apt, dnf or brew. Today, over 70 kubectl plugins are available on Krew.` It makes sense to offer OpenShift CLIs in the same way.
+    * With krew, artifacts will be packaged within an image.  Each cluster will have an index of CLI images.  Mirrored local registries will provide binaries to disconnected environments.  
+    * We will maintain an openshift-krew-index with supported artifacts. However, with `krew index add https://someother-third-party-index` 
+    we won't limit cluster-admins from adding their own index with whatever plugins they want.  That will be unsupported, though.
+
```

Comment:
Our IDE extensions do use `odo`, `tkn`, `kn`, `helm` underneath. Today, they either use the local copy if one is available or download their own copy if one is NOT available. `ConsoleCLIDownloads` allowed them to find these CLIs from the cluster. This proposes the CLIs to be shipped as images and also adds a dependency to `oc` and `krew` for their installation. This is a non-starter for IDEs.  IDEs require a https download option (no extraction from images) with no dependency to other tools for installation to be able to use this solution.

Reply:
noted.  `CLIs shipped as images` is meant to solve the `install disconnected` use with mirroring the images.  If this is not a requirement, (@smarterclayton @soltysh) shipping as images will not be necessary.  This proposal will add the ability to install the https download.  krew-index + CLI Custom Resource will hold the url for the CLI so that the current flow will still be available.